src/core/filter-from-graph.ts

```diff
@@ -1,9 +1,91 @@
 import { DepGraph, DepGraphInternal, NodeInfo, PkgInfo } from './types';
 import { DepGraphBuilder } from './builder';
+import { DepGraphImpl } from './dep-graph';
+import { eventLoopSpinner } from 'event-loop-spinner';
 
 type NodeId = string;
 type PkgName = string;
 
+/**
+ * Filter packages present in both graphs. Returns a induced subgraph of {@param graphB} with only packages
+ * that are not present in {@param graphA} or have a different version.
+ *
+ * @param graphA
+ * @param graphB
+ */
+export async function filterUnchangedPackages(
+  graphA: DepGraph,
+  graphB: DepGraph,
+): Promise<DepGraph> {
+  const depGraph = graphB as DepGraphInternal;
+
+  const graphAPackageIds = new Set(
+    graphA.getDepPkgs().map(DepGraphImpl.getPkgId),
+  );
+
+  const addedOrUpdatedPackages: PkgInfo[] = depGraph
+    .getDepPkgs()
+    .filter((pkg) => !graphAPackageIds.has(DepGraphImpl.getPkgId(pkg)));
+
+  const depGraphBuilder = new DepGraphBuilder(
+    depGraph.pkgManager,
+    depGraph.rootPkg,
+  );
+
+  const parentQueue: [parentId: NodeId, nodeId: NodeId][] = [];
+  for (const changedPackage of addedOrUpdatedPackages) {
+    for (const changedNodeId of depGraph.getPkgNodeIds(changedPackage)) {
+      //we add all nodes with new and changed packages to the new graph.
+      //a newly added node will also have its dependencies added here, since they are "new".
+      depGraphBuilder.addPkgNode(
+        depGraph.getNodePkg(changedNodeId),
+        changedNodeId,
+        getNodeInfo(depGraph, changedNodeId),
+      );
+
+      //Push all direct parents of the changed nodes to a queue to later build up a path to root from that node
+      for (const parentId of depGraph.getNodeParentsNodeIds(changedNodeId)) {
+        parentQueue.push([parentId, changedNodeId]);
+
+        if (eventLoopSpinner.isStarving()) {
+          await eventLoopSpinner.spin();
+        }
+      }
+    }
+  }
+
+  //add direct and transitive parents for the changed nodes
+  const visited = new Set([depGraph.rootNodeId]);
+
+  while (parentQueue.length > 0) {
```

Comment:
looks similar to filterPackagesFromGraph do we just call it?

Reply:
True, `filterPackagesFromGraph` could be used instead. I did not kind of by accident. Initially I wanted to keep all dependencies of a changed node, so it was not a good fit. But thinking on it more it was not worth to do, so I took that out. I still kept the implementation instead of calling `filterPackagesFromGraph` for (probably premature) optimization reasons. Calculating the packages to keep already give us the nodes to have in the graph so it's easier to just add them to the new graph and connect them and most changes to a graph are small, so iterating through the nodes to keep instead of all the ones to discard would be faster.